core/src/main/scala/com/lightbend/paradox/ParadoxProcessor.scala

```diff
@@ -78,25 +81,122 @@ class ParadoxProcessor(reader: Reader = new Reader, writer: Writer = new Writer)
+      errorCollector.logErrors(logger)
+      Left(s"Paradox failed with ${errorCollector.errorCount} errors")
+    } else Right(results)
+  }
+
+  /**
+   * Validate all mappings to build the site.
+   */
+  def validate(
+    mappings:         Seq[(File, String)],
+    allSiteFiles:     Seq[(File, String)],
+    groups:           Map[String, Seq[String]],
+    properties:       Map[String, String],
+    ignorePaths:      List[Regex],
+    validateAbsolute: Boolean,
+    logger:           ParadoxLogger): Int = {
+
+    val errorCollector = new ErrorCollector
+
+    val roots = parsePages(mappings, identity, properties, errorCollector)
+    val pages = Page.allPages(roots)
+    val globalPageMappings = rootPageMappings(roots)
+    val fullSite = allSiteFiles.map(_.swap).toMap
+
+    val linkCapturer = new LinkCapturer
+
+    @tailrec
+    def validate(location: Option[Location[Page]]): Unit = location match {
+      case Some(loc) =>
+        val page = loc.tree.label
+        val pageProperties = properties ++ page.properties.get
+        val currentMapping = Path.generateTargetFile(Path.relativeLocalPath(page.rootSrcPage, page.file.getPath), globalPageMappings)
+        val writerContext = Writer.Context(loc, pages, reader, writer, new PagedErrorContext(errorCollector, page),
+          logger, currentMapping, "", "", groups, pageProperties)
+        val serializer = linkCapturer.serializer(writerContext)
+        page.markdown.accept(serializer)
+        validate(loc.next)
+      case None => ()
+    }
+
+    roots.foreach { root =>
+      validate(Some(root.location))
+    }
+
+    def ignore(path: String) = ignorePaths.exists(_.pattern.matcher(path).matches())
+
+    linkCapturer.allLinks.foreach {
+      case CapturedLinkWithSources(CapturedAbsoluteLink(uri), sources) if validateAbsolute && !ignore(uri.toString) =>
+        validateExternalLink(uri, sources, errorCollector, logger)
+      case CapturedLinkWithSources(CapturedRelativeLink(path, fragment), sources) if !ignore(path) =>
+        fullSite.get(path) match {
+          case Some(file) =>
+            fragment.foreach { f =>
+              validateFragment(path, Source.fromFile(file, "UTF-8").mkString, f, sources, errorCollector)
+            }
+          case None =>
+            sources.foreach {
+              case (file, node) => errorCollector(s"Could not find path [$path] in site", file, node)
+            }
+
+        }
+      case _ =>
+      // Ignore
+    }
+
+    errorCollector.logErrors(logger)
+    errorCollector.errorCount
+  }
+
+  private def validateExternalLink(uri: URI, sources: List[(File, Node)], errorContext: ErrorContext, logger: ParadoxLogger) = {
+    logger.info(s"Validating external link: $uri")
+    val conn = uri.toURL.openConnection().asInstanceOf[HttpURLConnection]
+    conn.addRequestProperty("User-Agent", "Paradox Link Validator <https://github.com/lightbend/paradox>")
+    try {
+      if (conn.getResponseCode != 200) {
```

Comment:
Got several 301s and 302s on some old links. Do you think it should follow redirects when validating, or just require these links to be updated?

Reply:
It's nice UX to update redirected links and avoid the extra round trip. On the other hand, it can be annoying to have a hard failure on redirects, especially if it's running as part of pull request validation, and the redirected link has nothing to do with your change. IMHO it's better to run link validation only in cron-type builds rather than PR validation, unless you can be smart about only testing links that appear in the diff. In that case, knowing about redirects is a nice feature.

Reply:
Agree, better to know about the redirects, fail and update. And not run the external link validation in PR validation.

Reply:
Bonus UX on the redirect errors would be logging the new link as part of the error.

Reply:
Updated to report the location.

Reply:
Also, with regards to whether redirects should be a failure, the docs strongly recommend against running external link validation as part of CI (particularly for PR validation), so the intent is there to say that that use case is not the priority.

The HTTP spec actually says that anything that receives a 301 redirect should update any references to the URL that it has stored, so technically, by failing, we are enforcing the HTTP spec. And it's not just about avoiding the round trip - when a site starts enforcing https, it's better to link to the https version, though admittedly this is not as valuable as it once was, now that most dev websites force https. Also, URLs that 301 often eventually go dead, a 301 is sometimes used as early warning that something has moved and will be removed later. So for these reasons, I prefer to fail on 301s.

There are some cases where you might want to link to URLs that 301 (eg, linking to the latest version of the docs, on a site that 301s to the actual version), to support that we could make this configurable, but I'd say let's wait for people to request that. IMHO it's better to define an extref to the actual version of the docs, and use that, then you can update the version of the docs for that project that you link to with one configuration change. This avoids the tendency for links to the latest version of the docs going dead, links to actual versions tend to be much more stable.